core/src/test/java/org/apache/iceberg/hadoop/TestCachingCatalog.java

```diff
@@ -257,6 +264,44 @@ public void testCacheExpirationEagerlyRemovesMetadataTables() throws IOException
             catalog.cache().asMap().containsKey(metadataTable)));
   }
 
+  @Test
+  @Ignore("reproduces https://github.com/apache/iceberg/issues/3791")
+  public void testDeadlock() throws IOException, InterruptedException {
+    HadoopCatalog underlyingCatalog = hadoopCatalog();
+    TestableCachingCatalog catalog = TestableCachingCatalog.wrap(underlyingCatalog, Duration.ofSeconds(1), ticker);
+    Namespace namespace = Namespace.of("db", "ns1", "ns2");
+    int numThreads = 20;
+    for (int i = 0; i < numThreads; i++) {
+      TableIdentifier tableIdent = TableIdentifier.of(namespace, "tbl" + i);
+      catalog.createTable(tableIdent, SCHEMA, SPEC, ImmutableMap.of("key", "value"));
```

Comment:
Nit / being good citizens: Please make sure all of these tables get dropped at the end test! Maybe make a list of table names and then you can place a finally block or an `@After` function somehow.

Let me know if you'd like any help in sorting that out structurally etc 🙂

Reply:
done. I simply got the table names and dropped them at the end of this test. Maybe we can add a `@After` function that lists the tables and drops them all, but it's probably better to do it in a separate PR